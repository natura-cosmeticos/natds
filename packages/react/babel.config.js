const path = require('path')

const defaultAlias = {
  '@naturacosmeticos/natds-react': path.resolve(__dirname, './src/components')
}

module.exports = {
  plugins: [
    ['babel-plugin-module-resolver', { alias: defaultAlias }]
  ],
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    '@babel/preset-typescript',
    '@babel/preset-react'
  ]
}
