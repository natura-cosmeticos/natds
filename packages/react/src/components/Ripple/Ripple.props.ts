import { Color } from '@naturacosmeticos/natds-themes'

type RippleColors = keyof Pick<Color, 'primary' | 'secondary' | 'highlight'>

export interface RippleProps {
  /**
   * The element that will have the ripple effect applied
   */
  children: React.ReactNode
  /**
   * The color of ripple
   */
  color?: RippleColors
  /**
   * If `true`, the ripple effect is disabled
   */
  disabled: boolean
  /**
   *
   */
  fullWidth: boolean
  /**
   * If `true`, the ripple exceeds the element
   */
  hideOverflow?: boolean
}
