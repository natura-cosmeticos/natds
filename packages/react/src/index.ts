export * from './ThemeProvider'
export { useTheme, createUseStyles } from 'react-jss'

// Components
export * from './components/Avatar'
export * from './components/Badge'
export * from './components/Button'
export * from './components/Card'
export * from './components/Checkbox'
export * from './components/Counter'
export * from './components/Dialog'
export * from './components/Divider'
export * from './components/Icon'
export * from './components/IconButton'
export * from './components/Input'
export * from './components/Label'
export * from './components/Link'
export * from './components/ListItem'
export * from './components/Logo'
export * from './components/ProgressIndicator'
export * from './components/RadioButton'
export * from './components/Rating'
export * from './components/TextField'
