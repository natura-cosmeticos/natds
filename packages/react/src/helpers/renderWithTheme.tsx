import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
