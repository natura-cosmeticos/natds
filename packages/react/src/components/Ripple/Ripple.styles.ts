/* eslint-disable max-lines-per-function */
/* eslint-disable max-len */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { RippleProps } from './Ripple.props'

type RippleStyleProps = Required<Pick<RippleProps, 'color' | 'hideOverflow' | 'disabled' | 'fullWidth' | 'showHover' | 'animationDuration'>> &
{ size: number, mousePosition: { x: number, y: number }, isCentered: boolean }

const styles = createUseStyles((theme: Theme) => ({
  sharedRippleEffect: {
    content: '" "',
    height: '100%',
    left: 0,
    position: 'absolute',
    top: 0,
    width: '100%',
    zIndex: -1
  },
  wrapper: {
    flexBasis: ({ fullWidth }: RippleStyleProps) => fullWidth && '100%',
    cursor: ({ disabled }: RippleStyleProps) => (disabled ? 'default' : 'pointer'),
    display: ({ fullWidth }: RippleStyleProps) => (fullWidth ? 'block' : 'inline-block'),
    position: 'relative',
    alignSelf: ({ fullWidth }: RippleStyleProps) => (fullWidth ? 'stretch' : 'start'),
    outline: 'none',
    height: 'fit-content',
    '&:focus:after': {
      extend: 'sharedRippleEffect',
      backgroundColor: ({ color, disabled }: RippleStyleProps) => !disabled && theme.color[color],
      opacity: theme.opacity.mediumLow
    },
    '&:hover:after': {
      extend: 'sharedRippleEffect',
      backgroundColor: ({ color }: RippleStyleProps) => theme.color[color],
      opacity: ({ showHover, disabled }: RippleStyleProps) => (!disabled && showHover ? theme.opacity.mediumLow : 0)
    }
  },
  ripple: {
    backgroundColor: ({ color }: RippleStyleProps) => theme.color[color],
    borderRadius: '50%',
    height: ({ size }) => size,
    left: ({ mousePosition, isCentered }) => (isCentered ? '50%' : mousePosition.x),
    opacity: theme.opacity.transparent,
    position: 'absolute',
    top: ({ mousePosition, isCentered }) => (isCentered ? '50%' : mousePosition.y),
    transform: 'translate(-50%, -50%)',
    transition: ({ animationDuration }) => `scale opacity ${animationDuration}ms`,
    width: ({ size }) => size
  },
  rippleContainer: {
    inset: 0,
    overflow: ({ hideOverflow }: RippleStyleProps) => (hideOverflow && 'hidden'),
    position: 'absolute'
  },
  rippleActive: {
    animation: ({ animationDuration }) => `$ripple ${animationDuration}ms`,
    zIndex: 99
  },
  '@keyframes ripple': {
    '0%': {
      transform: 'translate(-50%, -50%) scale(0)'
    },
    '75%': {
      transform: 'translate(-50%, -50%) scale(1)',
      opacity: theme.opacity.low
    },
    '100%': {
      opacity: theme.opacity.transparent,
      zIndex: 0
    }
  }
}))

export default styles
