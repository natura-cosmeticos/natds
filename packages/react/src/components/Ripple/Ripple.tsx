import React, { useState } from 'react'
import styles from './Ripple.styles'
import { RippleProps } from './Ripple.props'

type Size = { width: number, height: number }
type MousePosition = { x: number, y: number }

export const getBiggestSide = ({ width, height }: Size): number => (width > height ? width : height)

const Ripple = ({
  animationDuration = 300,
  children,
  color = 'highlight',
  disabled = false,
  fullWidth = false,
  hideOverflow = true,
  isCentered = false
}: RippleProps): JSX.Element => {
  const [animation, setAnimation] = useState('')
  const [mousePosition, setMousePosition] = useState<MousePosition>({ x: 0, y: 0 })
  const [rippleSize, setRippleSize] = useState<Size>({ width: 0, height: 0 })

  const showRipple = (e: React.MouseEvent<HTMLDivElement>) => {
    if (!disabled) {
      setAnimation(rippleActive)
      setTimeout(() => setAnimation(''), animationDuration)
    }

    const {
      width, height, x, y
    } = e.currentTarget.getBoundingClientRect()

    const posX = e.pageX - x
    const posY = e.pageY - y

    setMousePosition({ x: posX, y: posY })
    setRippleSize({ width, height })
  }

  const size = getBiggestSide(rippleSize)

  const {
    ripple, rippleActive, rippleContainer, wrapper
  } = styles({
    animationDuration,
    color,
    disabled,
    fullWidth,
    hideOverflow,
    isCentered,
    mousePosition,
    size
  })

  return (
    <div className={wrapper} onClick={showRipple} data-testid="ripple-wrapper">
      <div className={rippleContainer}>
        <div className={`${ripple} ${animation}`} data-testid="ripple-animation" />
      </div>
      {children}
    </div>
  )
}

export default Ripple
