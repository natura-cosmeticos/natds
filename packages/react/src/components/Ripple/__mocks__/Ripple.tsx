import React from 'react'
import { RippleProps } from '../Ripple.props'

export const Ripple = ({
  children,
  ...props
}: RippleProps): JSX.Element => <div {...props}>{children}</div>

export default Ripple
