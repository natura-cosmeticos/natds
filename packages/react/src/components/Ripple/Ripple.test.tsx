import React from 'react'
import { act, fireEvent } from '@testing-library/react'
import renderWithTheme from '../../helpers/renderWithTheme'
import { Ripple } from '.'
import { getBiggestSide } from './Ripple'

jest.useFakeTimers()

const classWithRippleAnimation = 'ripple-0-2-2 ripple-d1-0-2-7 rippleActive-0-2-4 rippleActive-d3-0-2-9'

describe('Ripple', () => {
  it('should render correctly with children', () => {
    const { styles, component } = renderWithTheme(<Ripple> any children </Ripple>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
  it('should return the biggest side of container', () => {
    const container = { width: 214, height: 40 }
    const result = getBiggestSide(container)

    expect(result).toBe(214)
  })
  it('should add ripple animation class', async () => {
    const { component } = renderWithTheme(<Ripple> any children </Ripple>)

    const wrapper = component.getByTestId('ripple-wrapper')
    const ripleAnimation = component.getByTestId('ripple-animation')

    expect(ripleAnimation).not.toHaveClass(classWithRippleAnimation)
    fireEvent.click(wrapper)
    expect(ripleAnimation).toHaveClass(classWithRippleAnimation)

    act(() => jest.advanceTimersByTime(1000))
    expect(ripleAnimation).not.toHaveClass(classWithRippleAnimation)
  })
  it('should not add ripple animation class when Ripple is disabled', async () => {
    const { component } = renderWithTheme(<Ripple disabled> any children </Ripple>)

    const wrapper = component.getByTestId('ripple-wrapper')
    const ripleAnimation = component.getByTestId('ripple-animation')

    expect(ripleAnimation).not.toHaveClass(classWithRippleAnimation)
    fireEvent.click(wrapper)
    expect(ripleAnimation).not.toHaveClass(classWithRippleAnimation)
  })
})
