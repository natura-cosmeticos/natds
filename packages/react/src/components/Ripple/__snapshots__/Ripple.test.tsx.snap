// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple should render correctly with children 1`] = `
Array [
  ".wrapper-0-2-1 {
  cursor: pointer;
  position: relative;
  align-self: start;
}
.ripple-0-2-2 {
  opacity: 0;
  position: absolute;
  transform: translate(-50%, -50%);
  border-radius: 50%;
}
.rippleContainer-0-2-3 {
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
}
.rippleActive-0-2-4 {
  z-index: 1;
}
@keyframes keyframes-ripple-0-2-5 {
  0% {
    transform: translate(-50%, -50%) scale(0);
  }
  75% {
    opacity: 0.12;
    transform: translate(-50%, -50%) scale(1);
  }
  100% {
    opacity: 0;
    z-index: 0;
  }
}
  .wrapper-d0-0-2-6 {
    display: inline-block;
  }
  .ripple-d1-0-2-7 {
    background-color: #000000;
    height: 0;
    left: 0;
    top: 0;
    transition: scale opacity 300ms;
    width: 0;
  }
  .rippleContainer-d2-0-2-8 {
    overflow: hidden;
  }
  .rippleActive-d3-0-2-9 {
    animation: keyframes-ripple-0-2-5 300ms;
  }",
  <div>
    <div
      class="wrapper-0-2-1 wrapper-d0-0-2-6"
      data-testid="ripple-wrapper"
    >
      <div
        class="rippleContainer-0-2-3 rippleContainer-d2-0-2-8"
      >
        <div
          class="ripple-0-2-2 ripple-d1-0-2-7 "
          data-testid="ripple-animation"
        />
      </div>
       any children 
    </div>
  </div>,
]
`;
